upstairs/src/volume.rs

```diff
@@ -1238,14 +1234,130 @@ impl Volume {
+                    // to a VCR where there was not one originally.
+                    crucible_bail!(
+                        ReplaceRequestInvalid,
+                        "Unexpected read_only_parent on new VCR"
+                    )
+                }
+
+                if o_read_only_parent != n_read_only_parent {
+                    // To get here, both read only parents are populated.
+
+                    let old_rop = o_read_only_parent.unwrap();
+                    let new_rop = n_read_only_parent.unwrap();
+                    // So, the ROPs are different.  Now let's be sure
+                    // they are different in only a target.
+                    Self::compare_vcr_region_for_replacement(&old_rop, &new_rop)
+                } else {
+                    // If the read_only_parents are the same, then this is
+                    // technically an error (from the ROP replacement point of
+                    // view) as it means the targets and generation numbers
+                    // are the same as what we have.
+                    Err(CrucibleError::ReplaceRequestInvalid(
+                        "read_only_parent generation numbers match".to_string(),
+                    ))
+                }
+            } else {
+                // New read_only_parent is none, so the read_only_parents can't
+                // be considered for replacement.
+                Err(CrucibleError::ReplaceRequestInvalid(
+                    "read_only_parent does not exist in replacement"
+                        .to_string(),
+                ))
+            }
+        };
+
+        // Determine the sub_volume difference if there is any.
+        let sub_volume_result = Self::compare_vcr_region_for_replacement(
+            &o_sub_volumes[0],
+            &n_sub_volumes[0],
+        );
+        info!(
+            log,
+            "sub_volume_result is: {:?}, ROP:{:?}",
+            sub_volume_result,
+            rop_result
+        );
+
+        // So, now time for the truth.
+        // We use the same function to compare both the region in the
+        // sub_volumes and the region in the read_only_parent.  Only one
+        // can be different, so we walk the cases.
+        match sub_volume_result {
+            Ok(Some((old, new))) => {
+                // The sub_volume had a valid difference, if the
+                // read_only_parent also had a difference, then that is an
+                // error, otherwise, we return what the sub_volume found.
+                match rop_result {
+                    Ok(Some((_, _))) => {
+                        // Both have changes, this is no good
+                        crucible_bail!(
+                            ReplaceRequestInvalid,
+                            "Both sub_volume and read_only_parent have changes"
+                        )
+                    }
+                    _ => Ok(Some((old, new))),
```

Comment:
should this match on `rop_result` being `Ok(None)` ?

Reply:
Yes... I believe the code as is would ignore an error in `rop_result` if `sub_volume_result` is
`Some()` and instead of returning error it returns what the `SubVolume` found.  This is tricky
because, when comparing two VCRs,  we can have only one of these:
1) A valid ROP difference only, SubVolumes are the same.
2) The ROPs are identical and the new SubVolume is different.
3) The Old has a ROP and the new does not, and the SubVolume is different.

I'll ponder this..

Reply:
I've refactored a bunch of this and it now should cover all cases as well as
(I hope) make better sense.